File: ios/Sources/MoQKit/Subscribe/MediaTrack.swift
import Foundation
import MoqFFI

// MARK: - Media Container

/// Container format used by a MoQ media track.
///
/// Catalog-advertised tracks provide this automatically via ``AudioTrackInfo`` and
/// ``VideoTrackInfo``. Advanced callers can provide it directly when subscribing to a known
/// media track that is not advertised in the catalog.
public enum MediaContainer: Sendable, Equatable, Hashable {
    /// Legacy MoQ media container.
    case legacy
    /// CMAF/fMP4 media container with the initialization segment bytes.
    case cmaf(initializationData: Data)
    /// LOC media container.
    case loc

    init(_ raw: Container) {
        switch raw {
        case .legacy:
            self = .legacy
        case .cmaf(let initializationData):
            self = .cmaf(initializationData: initializationData)
        case .loc:
            self = .loc
        }
    }

    var rawContainer: Container {
        switch self {
        case .legacy:
            return .legacy
        case .cmaf(let initializationData):
            return .cmaf(init: initializationData)
        case .loc:
            return .loc
        }
    }
}

// MARK: - Media Track Request

/// Parameters needed to subscribe to a MoQ media track.
///
/// Use this when subscribing to media by name from a ``Broadcast`` without relying on catalog
/// metadata. When multiple consumers subscribe to the same track, the first subscriber creates
/// the shared upstream subscription and chooses its ``targetBuffering``.
public struct MediaTrackRequest: Sendable, Equatable {
    /// Track name on the broadcast.
    public let name: String
    /// Track container format.
    public let container: MediaContainer
    /// Target live buffering depth for the upstream media subscription.
    public let targetBuffering: Duration

    public init(
        name: String,
        container: MediaContainer,
        targetBuffering: Duration = .milliseconds(100)
    ) {
        self.name = name
        self.container = container
        self.targetBuffering = targetBuffering
    }

    init(track: AudioTrackInfo, targetBuffering: Duration) {
        self.init(
            name: track.name,
            container: MediaContainer(track.rawConfig.container),
            targetBuffering: targetBuffering
        )
    }

    init(track: VideoTrackInfo, targetBuffering: Duration) {
        self.init(
            name: track.name,
            container: MediaContainer(track.rawConfig.container),
            targetBuffering: targetBuffering
        )
    }
}

// MARK: - MediaFrame

/// A single compressed media frame received from the relay.
///
/// Frames are delivered in presentation order on ``MediaTrack/frames``.
/// Most consumers should use ``Player`` rather than working with frames directly.
public struct MediaFrame: Sendable {
    /// The raw compressed payload bytes (e.g. an H.264 NAL unit, an AAC packet).
    public let payload: Data
    /// Presentation timestamp in microseconds, relative to the stream origin.
    public let timestampUs: UInt64
    /// Whether this frame is a keyframe (sync point). Keyframes can be decoded independently
    /// without prior frames.
    public let keyframe: Bool

    init(_ raw: MoqFrame) {
        self.payload = raw.payload
        self.timestampUs = raw.timestampUs
        self.keyframe = raw.keyframe
    }

    init(payload: Data, timestampUs: UInt64, keyframe: Bool) {
        self.payload = payload
        self.timestampUs = timestampUs
        self.keyframe = keyframe
    }
}

// MARK: - Media Track Options

/// Buffering behavior for frames emitted by a ``MediaTrack``.
public enum MediaTrackBufferingPolicy: Sendable, Equatable {
    /// Buffers every frame until the consumer reads it.
    case unbounded
    /// Keeps only the newest `limit` frames when the consumer falls behind.
    ///
    /// Non-positive limits are treated as `1`.
    case bufferingNewest(Int)

    var streamPolicy: AsyncThrowingStream<MediaFrame, Error>.Continuation.BufferingPolicy {
        switch self {
        case .unbounded:
            return .unbounded
        case .bufferingNewest(let limit):
            return .bufferingNewest(max(1, limit))
        }
    }
}

/// Options for subscribing to a compressed media track.
public struct MediaTrackOptions: Sendable, Equatable {
    /// Buffering behavior for delivered frames.
    public let bufferingPolicy: MediaTrackBufferingPolicy

    public init(
        bufferingPolicy: MediaTrackBufferingPolicy = .unbounded
    ) {
        self.bufferingPolicy = bufferingPolicy
    }
}

// MARK: - Track State

/// The lifecycle state of a ``MediaTrack``.
public enum MediaTrackState: Sendable, Equatable {
    /// Subscribed to the track but no frames have arrived yet.
    case idle
    /// Frames are arriving and being emitted on ``MediaTrack/frames``.
    case active
    /// The track ended normally (the remote sender closed it).
    case closed
    /// The track ended with an error. The associated string contains a human-readable description.
    case error(String)
}

// MARK: - Media Track

/// Advanced low-level subscription to one MoQ media track.
///
/// `MediaTrack` surfaces raw ``MediaFrame`` values from the relay so you can plug them into
/// your own decoder or processing pipeline. In most apps, use ``Player`` instead, because
/// it manages subscriptions, decoding, buffering, and rendering for you.
///
/// ``frames`` is an `AsyncThrowingStream` that completes when the track ends or throws when
/// the underlying media subscription fails. ``state`` mirrors the same lifecycle for consumers
/// that prefer state events.
public final class MediaTrack: @unchecked Sendable {
    /// A stream of raw media frames as they arrive from the relay.
    public let frames: AsyncThrowingStream<MediaFrame, Error>
    /// A stream of ``MediaTrackState`` transitions. Always yields `.idle` as its first element.
    public let state: AsyncStream<MediaTrackState>

    private let media: MediaFrameStream
    private let framesContinuation: AsyncThrowingStream<MediaFrame, Error>.Continuation
    private let stateContinuation: AsyncStream<MediaTrackState>.Continuation
    private var readTask: Task<Void, Never>?

    init(
        media: MediaFrameStream,
        options: MediaTrackOptions = MediaTrackOptions()
    ) {
        self.media = media

        var framesCont: AsyncThrowingStream<MediaFrame, Error>.Continuation!
        let framesStream = AsyncThrowingStream<MediaFrame, Error>(
            bufferingPolicy: options.bufferingPolicy.streamPolicy
        ) { framesCont = $0 }

        var stateCont: AsyncStream<MediaTrackState>.Continuation!
        let stateStream = AsyncStream<MediaTrackState> { stateCont = $0 }

        self.frames = framesStream
        self.state = stateStream
        self.framesContinuation = framesCont
        self.stateContinuation = stateCont

        stateCont.yield(.idle)

        readTask = Task.detached {
            var isFirstFrame = true
            var didFinishFrames = false
            defer {
                if !didFinishFrames {
                    framesCont.finish()
                }
                stateCont.finish()
            }

            do {
                for try await frame in media.frames {
                    guard !Task.isCancelled else { return }
                    if isFirstFrame {
                        stateCont.yield(.active)
                        isFirstFrame = false
                    }
                    framesCont.yield(frame)
                }

                if !Task.isCancelled {
                    stateCont.yield(.closed)
                }
            } catch {
                if !Task.isCancelled {
                    stateCont.yield(.error(error.localizedDescription))
                    framesCont.finish(throwing: error)
                    didFinishFrames = true
                }
            }
        }
    }

    /// Cancels the track subscription and completes both ``frames`` and ``state`` streams.
    ///
    /// Safe to call multiple times.
    public func close() {
        media.close()
        readTask?.cancel()
        readTask = nil
        // Do not yield .closed here — the read task's defer handles that on a normal
        // end, and double-yielding would emit two .closed events to state consumers.
        stateContinuation.finish()
        framesContinuation.finish()
    }

    deinit {
        media.close()
        readTask?.cancel()
        stateContinuation.finish()
        framesContinuation.finish()
    }
}

File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/subscribe/MediaTrack.kt
package com.swmansion.moqkit.subscribe

import com.swmansion.moqkit.subscribe.internal.MediaFrameEvent
import com.swmansion.moqkit.subscribe.internal.MediaFrameStream
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.asStateFlow
import uniffi.moq.Container
import uniffi.moq.MoqFrame
import java.time.Duration

/**
 * Container format used by a MoQ media track.
 *
 * Catalog-advertised tracks provide this automatically through [AudioTrackInfo] and
 * [VideoTrackInfo]. Advanced callers can provide it directly when subscribing to a known
 * media track that is not advertised in the catalog.
 */
sealed class MediaContainer {
    /** Legacy MoQ media container. */
    object Legacy : MediaContainer()

    /** LOC media container. */
    object Loc : MediaContainer()

    /**
     * CMAF/fMP4 media container with initialization segment bytes.
     */
    class Cmaf(initializationData: ByteArray) : MediaContainer() {
        val initializationData: ByteArray = initializationData.copyOf()

        override fun equals(other: Any?): Boolean =
            other is Cmaf && initializationData.contentEquals(other.initializationData)

        override fun hashCode(): Int = initializationData.contentHashCode()

        override fun toString(): String =
            "Cmaf(initializationData=${initializationData.contentToString()})"
    }

    internal fun toRawContainer(): Container = when (this) {
        Legacy -> Container.Legacy
        Loc -> Container.Loc
        is Cmaf -> Container.Cmaf(initializationData.copyOf())
    }

    internal companion object {
        fun fromRaw(container: Container): MediaContainer = when (container) {
            Container.Legacy -> Legacy
            Container.Loc -> Loc
            is Container.Cmaf -> Cmaf(container.init)
        }
    }
}

/**
 * Parameters needed to subscribe to a MoQ media track.
 *
 * When multiple consumers subscribe to the same track on the same broadcast, the first
 * subscriber creates the shared upstream subscription and chooses [targetBuffering].
 */
data class MediaTrackRequest(
    /** Track name on the broadcast. */
    val name: String,
    /** Track container format. */
    val container: MediaContainer,
    /** Target live buffering depth for the upstream media subscription. */
    val targetBuffering: Duration = Duration.ofMillis(100),
) {
    internal constructor(track: AudioTrackInfo, targetBuffering: Duration) : this(
        name = track.name,
        container = MediaContainer.fromRaw(track.rawConfig.container),
        targetBuffering = targetBuffering,
    )

    internal constructor(track: VideoTrackInfo, targetBuffering: Duration) : this(
        name = track.name,
        container = MediaContainer.fromRaw(track.rawConfig.container),
        targetBuffering = targetBuffering,
    )
}

/**
 * Buffering behavior for frames emitted by a [MediaTrack].
 */
sealed class MediaTrackBufferingPolicy {
    /** Buffers every frame until the consumer reads it. */
    object Unbounded : MediaTrackBufferingPolicy()

    /**
     * Keeps only the newest [limit] frames when the consumer falls behind.
     *
     * Non-positive limits are treated as `1`.
     */
    data class BufferingNewest(val limit: Int) : MediaTrackBufferingPolicy()
}

/**
 * Options for subscribing to a compressed media track.
 */
data class MediaTrackOptions(
    /** Buffering behavior for delivered frames. */
    val bufferingPolicy: MediaTrackBufferingPolicy = MediaTrackBufferingPolicy.Unbounded,
)

/**
 * A single compressed media frame received from the relay.
 */
class MediaFrame(
    /** Raw compressed payload bytes. */
    val payload: ByteArray,
    /** Presentation timestamp in microseconds, relative to the stream origin. */
    val timestampUs: Long,
    /** Whether this frame is a keyframe or sync point. */
    val keyframe: Boolean,
) {
    internal constructor(raw: MoqFrame) : this(
        payload = raw.payload,
        timestampUs = raw.timestampUs.toLong(),
        keyframe = raw.keyframe,
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is MediaFrame) return false
        return payload.contentEquals(other.payload) &&
            timestampUs == other.timestampUs &&
            keyframe == other.keyframe
    }

    override fun hashCode(): Int {
        var result = payload.contentHashCode()
        result = 31 * result + timestampUs.hashCode()
        result = 31 * result + keyframe.hashCode()
        return result
    }

    override fun toString(): String =
        "MediaFrame(payload=${payload.contentToString()}, timestampUs=$timestampUs, keyframe=$keyframe)"
}

/**
 * Lifecycle state of a [MediaTrack].
 */
sealed class MediaTrackState {
    /** Subscribed to the track but no frames have arrived yet. */
    object Idle : MediaTrackState()

    /** Frames are arriving and being emitted on [MediaTrack.frames]. */
    object Active : MediaTrackState()

    /** The track ended normally or was closed by the caller. */
    object Closed : MediaTrackState()

    /**
     * The track ended with an error.
     *
     * @property message Human-readable error description.
     */
    data class Error(val message: String) : MediaTrackState()
}

/**
 * Advanced low-level subscription to one MoQ media track.
 *
 * [frames] emits raw compressed [MediaFrame] values. Most applications should use [Player]
 * or [AudioDataStream], which manage decoding and playback or processing.
 */
class MediaTrack internal constructor(
    private val media: MediaFrameStream,
    private val onClose: (() -> Unit)? = null,
) : AutoCloseable {
    private val lock = Any()
    private val _state = MutableStateFlow<MediaTrackState>(MediaTrackState.Idle)
    private var closed = false

    internal val events: Flow<MediaFrameEvent> = flow {
        var firstFrame = true
        try {
            media.events.collect { event ->
                if (firstFrame && event is MediaFrameEvent.Frame) {
                    firstFrame = false
                    _state.value = MediaTrackState.Active
                }
                emit(event)
            }
            _state.value = MediaTrackState.Closed
        } catch (t: Throwable) {
            if (!isClosed) {
                _state.value = MediaTrackState.Error(t.message ?: t::class.java.simpleName)
            }
            throw t
        } finally {
            close()
        }
    }

    /** A stream of raw media frames as they arrive from the relay. */
    val frames: Flow<MediaFrame> = flow {
        events.collect { event ->
            if (event is MediaFrameEvent.Frame) emit(event.frame)
        }
    }

    /** Current lifecycle state for this track. */
    val state: StateFlow<MediaTrackState> = _state.asStateFlow()

    /** Whether this subscription has been closed. */
    val isClosed: Boolean
        get() = synchronized(lock) { closed }

    /** Cancels the track subscription and completes [frames]. */
    override fun close() {
        val shouldClose = synchronized(lock) {
            if (closed) {
                false
            } else {
                closed = true
                true
            }
        }
        if (!shouldClose) return

        media.close()
        _state.value = MediaTrackState.Closed
        onClose?.invoke()
    }
}
